### DISCUSSION TITLE
Shares function works but how do i share using my External IP or DNS

### DISCUSSION THREAD

**Discussion by bembudo** [ampache/ampache#3584] [General]
### Describe the bug
Shares function works but how do i share using my External IP or DNS ? 

### Expected behavior
I would like to create a share and have my external ip or DNS show up in the shared URL

### Screenshots
![Screenshot](https://user-images.githubusercontent.com/74638754/110186718-231c3a00-7de4-11eb-8371-588d61d58b0a.jpg)

## Environment

* Ampache version:
* Web server + version:

### Client type

* Web
* Ampache

**Comment by SePHTaN** [comment]
Hi bembudo,

I haven't tested shares functionaltity of ampache by myself. 
So must admit that my following suggestions are assumptions that they work as i expect though.

Simpelst Way : 
Find out your external ip using sites like https://www.wieistmeineip.de/
Replace the internal ip in the iframe by your external one by hand.
The qrcode wont be usable.

Medium complicated Way : 
If you use an ddns provider like me you could tweak your hosts file with that information and let it point your ddns-name point to your server.
If you use linux this could be (Debian/ubuntu based), example below is how mine would look
```
/etc/hosts
127.0.0.1      localhost
# exchange the fqdn with yours from your ddns-provider
192.168.3.8    ampache.sandmann.dd-dns.de
```
For Win10 this would be C:\Windows\System32\drivers\etc\hosts
Has examples inside and is self explaining.

Now you can open ampache from inside your network as if you are outside.
This should give you a qrcode and htmlcode to embed with fqdn (if ampache shares work as i assume)

The complicated way : 
Set up your own DNS server with a zone file for your ddns-domain pointing to the internal ip from your webserver.
In principle the same as tweaking the host file but a DNS Server gives more abilities in many ways.

I would suggest that you give the medium complicated way a try.
It's easy to do and undo
Not to forget when you have changed your hosts file then you have to browse ampache with your FQDN and with your internal IP.
This means http://yourdyndns.domain:8080
Instead of  http://192.168.2.21:8080

**Comment by bembudo** [comment]
Hi SePHTaN

thank you for this, but these look like hacks to achieve functionality, i thought there was an easy way through Ampache configuration. Would'nt it make sense to have this ability since the share works for both internal and external ?

**Comment by SePHTaN** [comment]
I don't would call it a hack.

Simple comparison : try to to open a door from outside when you are inside. Not easy isn't it.
The problem for ampache is comparable to do this.

One question i do have, how do you acces ampache from the internet ? By IP or by a FullQualifiedDomainName ?
If you do it by IP i would suggest trying one of the free DynDNS Providers, there are many. 
Most Routers have an option to update the ip the FQDN points to.

If you access ampache by ip from inside and ouside your network it could be achievable.
Ampache would have to do the same as i suggested with "Simplest way". Get the ip from an service like the one i mentioned.
I don't know if any of them supplies an api for doing such things, if not one has to grab the ip from the html code.
When they change the layout of their webside most certainly this functionality is broken afterwards and needs to be adjusted to their changed html code.

**Comment by wagnered** [comment]
@bembudo and @SePHTaN ,

Also the router must be set up for "port forwarding".  or depending on the brand of router expose a computer to the outside world(dangerous).  either way,  the data from the outside world has to be translated from the public address to the local(internal) address of the computer.  And the computer being exposed should have a static internal address.

**Comment by bembudo** [comment]
I tried the hosts file solution on my Ubuntu machine, this does not work. I do have a Dynamic DNS name with [Freedns](https://freedns.afraid.org/) and I point the DNS name to my external ip which allows me to access Ampache externally, of course port forwarding is setup i route this traffic through port 8080 as displayed in my screen shot above.

@SePHTaN I dont know what you mean by iFrame ? first time hearing this term 

> Replace the internal ip in the iframe by your external one by hand.

I just thought there could be an option that says - "Share using External address" I dont care for the QRC Code

**Comment by bembudo** [comment]
if i manually modify my url and replace my internal address with my DNS name it works. But i shouldnt have to do this each time I share

**Comment by SePHTaN** [comment]
> I dont know what you mean by iFrame ? first time hearing this term

In the screenshot you postet is the code beneath the qrlogo if you want to embed the share into a website.
Technically this is done with an iframe and the code there clearly says:
`<iframe style="width:630px; height:75px;" src="http://192.168.2.21:8080/share.php?id=6&secret=secret&embed=true" ></iframe>`
Explanation of my suggestion:
If you browse the internet lets say _https://google.com_ your Computer always opens a connection to an IP address for google this could be (among many others) 142.250.185.142 you can try by just open https://142.250.185.142 in your browser you will get to google.com search site.
Normally your computer get the ip address for google.com from a DNS Server, but you can override this by giving this name-resolution by your hosts file.
This is btw. one way adware blockers do block ad, they make hosts entrys for known adware servers FQDN in your hostsfile pointing to 127.0.01 your localhost. This does work because your PC normally first looks into the hostsfile before asking any DNS servers.
In my network setup i have an homeserver with sambaserver and dnsserver as with several virtualmachines one running my webstuff like ampache and one a firewall application.
My normal PC and other internet capable devices are on the 192.168.1.0 network, the webserver is on 192.168.3.0 network.
My Firewall forwards port 80 & 443 (http/https) to the webserver network.
My DNS Server has a zone with my dyndnsdomain (ampache.sandmann.dd-dns.de) pointing to the ip of my webserver.
The result is the same as when i would put the ip in a hosts file.
So for me it makes no difference browsing my webserver with my smartphone wether i am inside my network (over wifi) or outside in the internet. I always browse https://ampache.sandmann.dd-dns.de.
The difference to the hostfile is the hostsfile is only for the PC the hostsfile is from a DNS server gives name/ip resolution to all devices in the internal network.

If you are on win10 there could be antivirus or win10 itself preventing the change of the hostsfile.
I know that windows protects the hostsfile because you can do malicous things with it like redirecting popular sites like google facebook or other to malware dropping websites.
So it might make sense to check wether your changes to the hostsfile were persistent.

**Comment by bembudo** [comment]
Im sorry SePHaN, as mentioned above the hosts file hack does not  work and im not looking to embed anything into a website so your workaround is simply replaying the analogy you provided of the door for me.

I just need a straight answer, can this posting be converted to  a feature request and is it possible to share a link using external IP or DNS ?

**Comment by wagnered** [comment]
I took the privilege to create a new feature request (#2807) and in it, referred to this issue.

**Comment by bembudo** [comment]
Thank you wagnered Greatly appreciated

**Comment by kuzi-moto** [comment]
@bembudo, I believe what you're asking is outside of the scope of Ampache. The shares are accessible from outside by default, it depends on configuration of your environment which Ampache has no control.

The shares will be created using the host section of the current URL of Ampache. For example https://192.168.1.2/home.php, the host is '192.168.1.2'. Assuming port forwarding is enabled correctly, anyone outside your network can access via the External IP of your network. To get a nice URL instead of an IP a dynamic DNS service can be used. However this will not work inside of the network as most routers do not understand how to handle requests from inside trying to access the external IP.

This can be solved by editing the hosts file of the machine trying to access Ampache, by entering a line such as `my.freedns.org 192.168.1.2` where the IP address is the internal IP address of the Ampache server. This overrides the address your computer would otherwise get from it's currently configured DNS server.

Just a note, some browsers have started to implement "secure" DNS and ignore the local hosts file and the DNS configured for your computer. You will need to disable this if it's the case otherwise setting things in the hosts doesn't work.

I think most/all of this was already said, hopefully this makes it a little more clear. I think the only thing Ampache could do is to provide a configuration option to override the host when creating shared URLs so you don't have to change it manually yourself.

**Comment by SePHTaN** [comment]
@kuzi-moto Thanks for clarifing this in a better way than my  english knowedge allowed.
The usecase with ampache and the almost impossibility to get this working with double nat is what brought me the hard way in learning how to setup a DNS Server solely for internal purposes.

Otherwise i can understand that not everybody wants/can go that way.
Before i mentioned that i don't belive that there exist external ip-checker offering an easy to use api, but a quick check through google told me that i am wrong : 
[A Simple Public IP Address API](https://www.ipify.org/)
With that api it could easily be achieved to have the external ip in the shared qrcode/URL.
The external IP should only be used when ampache isn't browsed per FQDN.
@bembudo The so created share would also be only accessible from outside

**Comment by lachlan-00** [comment]
just set a http_host in the config file
```
http_host = "external.hostname.com"
```

if you have a proxy, set the local_web_path as well
```
local_web_path = "http://192.168.1.1:8080"
```

**Comment by bembudo** [comment]
Okay so I managed to get this to work although I hope my feature request could be considered. Here is what i did going by suggestion.

I updated http_host in Ampache.cfg.php file to my dns host name
edited my hosts file in Ubuntu 18.04 to 192.168.2.21    <external.hostname.com>

now when i share, I see my external hostname rather than my internal ip which is good. But this limits my internal access to ampache web to the machine with the modified hosts file. No other computer in my local area network can access Ampache web.

I presume a proxy could fix this ? as @lachlan-00  mentions above

Lots of work just to share using an external address

**Comment by SePHTaN** [comment]
Only partially right as lachlan said:
`http_host = "external.hostname.com"`   
this your hostname for external access your external dyndns domain, with tweaking your hostsfile also from inside reacheable with this "external.hostname.com".

local_web_path = "http://192.168.2.21:8080"
should also give you the ability to reach your server by ip as before thats why its called LOCAL web_path, but shares qrcode and iframe then are as before with your internal ip and thus not really shareable for people accessing from outside.
but it should work for from inside

**Comment by bembudo** [comment]
thanks SePhTaN yes my local_web_path is http://192.168.2.21:8080 however not accessible from any other computer other than my ubuntu machine where ampache is configured.

Ill play around with it some more, if you have any additional suggestions or thoughts id be happy to hear them
